sql/backfill.go

```diff
@@ -80,3 +82,138 @@ func (p *planner) makeBackfillBatch(tableName *parser.QualifiedName, tableDesc *
+		tableDesc := &TableDescriptor{}
+		if err := p.getDescriptor(tableKey{desc.ParentID, tableName}, tableDesc); err != nil {
+			return err
+		}
+		if len(tableDesc.Mutations) == 0 {
+			return nil
+		}
+		log.Info("apply mutation")
+		mutation := tableDesc.Mutations[0]
+		tableDesc.Mutations = tableDesc.Mutations[1:]
+		if err := tableDesc.applyMutation(*mutation); err != nil {
+			return err
+		}
+		if indexDesc := mutation.GetAddIndex(); indexDesc != nil {
+			log.Infof("Attempt to Add index: %s", indexDesc.Name)
+			// `indexDesc` changed on us when we called `tableDesc.AllocateIDs()`.
+			if mutation.Primary {
+				indexDesc = &tableDesc.PrimaryIndex
+			} else {
+				indexDesc = &tableDesc.Indexes[len(tableDesc.Indexes)-1]
+			}
+
+			name := &parser.QualifiedName{Base: parser.Name(tableName)}
+			if err := name.NormalizeTableName(dbName); err != nil {
+				return err
+			}
+
+			b, err := p.makeBackfillBatch(name, tableDesc, *indexDesc)
+			if err != nil {
+				return err
+			}
+
+			b.Put(MakeDescMetadataKey(tableDesc.GetID()), tableDesc)
+			p.txn.SetSystemDBTrigger()
+
+			if err := p.txn.Run(&b); err != nil {
+				return convertBatchError(tableDesc, b, err)
+			}
+			log.Infof("Added index %s", indexDesc.Name)
+			return nil
+		}
+
+		if columnDesc := mutation.GetAddColumn(); columnDesc != nil {
+			log.Infof("Attempt to Add column: %s", columnDesc.Name)
+			name := &parser.QualifiedName{Base: parser.Name(tableName)}
+			if err := name.NormalizeTableName(dbName); err != nil {
+				return err
+			}
+
+			b, err := p.makeBackfillBatch(name, tableDesc)
+			if err != nil {
+				return err
+			}
+
+			b.Put(MakeDescMetadataKey(tableDesc.GetID()), tableDesc)
+			p.txn.SetSystemDBTrigger()
+
+			if err := p.txn.Run(&b); err != nil {
+				return convertBatchError(tableDesc, b, err)
+			}
+			log.Infof("Added column %s", columnDesc.Name)
+
+			return nil
+		}
+
+		if indexDesc := mutation.GetDropIndex(); indexDesc != nil {
+			log.Infof("Attempt to drop index %s", indexDesc.Name)
+
+			// delete index
+			indexPrefix := MakeIndexKeyPrefix(tableDesc.ID, indexDesc.ID)
+
+			// Drop the index.
+			indexStartKey := roachpb.Key(indexPrefix)
+			indexEndKey := indexStartKey.PrefixEnd()
+			if log.V(2) {
+				log.Infof("DelRange %s - %s", prettyKey(indexStartKey, 0), prettyKey(indexEndKey, 0))
+			}
+			b := client.Batch{}
+
+			b.DelRange(indexStartKey, indexEndKey)
+
+			descKey := MakeDescMetadataKey(tableDesc.GetID())
+			if err := tableDesc.Validate(); err != nil {
+				return err
+			}
+			b.Put(descKey, tableDesc)
+			p.txn.SetSystemDBTrigger()
+
+			if err := p.txn.Run(&b); err != nil {
+				return err
+			}
+			log.Infof("Dropped index %s", indexDesc.Name)
+			return nil
+		}
+		if columnDesc := mutation.GetDropColumn(); columnDesc != nil {
+			panic("Not Implemented")
+		}
+		panic("can't get here")
+	})
+	if err != nil {
+		if _, ok := err.(errUniquenessConstraintViolation); ok {
+			// irrecoverable error; delete all future mutations.
```

Comment:
There are more irrecoverable errors. Perhaps I should run a retry loop and declare the mutation as irrecoverable after a few attempts.